File: src/main/java/com/allaboutjava/chap09/Tutorial_02_Non_Blocking_Future_Result.java
package com.allaboutjava.chap09;

import io.vavr.concurrent.Future;
import io.vavr.control.Try;

import java.util.function.Function;

import static io.vavr.API.*;
import static io.vavr.Patterns.$Failure;
import static io.vavr.Patterns.$Success;
import static io.vavr.Predicates.instanceOf;

public class Tutorial_02_Non_Blocking_Future_Result {

    public static void main(String[] args) throws InterruptedException {

        println("Step 1: Define a method which returns a Future");
        Function<String, Future<Integer>> donutStock = __ -> Future.of(() -> {
            // assume some long running database operation
            Thread.sleep(500);
            println("checking donut stock");
            return 10;
        });

        println("\nStep 2: Call method which returns a Future");
        /**donutStock.apply("vanilla donut")
         .onSuccess(stock -> printf("Stock for vanilla donut = %s", stock))
         .onFailure(e -> printf("Failed to find vanilla donut stock, exception = %e"))
         .get();*/
        donutStock.apply("vanilla donut")
                .onComplete(response ->
                        Match(response).of(
                                Case($Success($(instanceOf(Integer.class))), stock -> run(() -> printf("Stock for vanilla donut = %s", stock))),
                                Case($Failure($()), e -> run(e::printStackTrace))
                        ));
        Thread.sleep(3000);
    }
}

File: src/main/scala/com/allaboutscala/chap09/Tutorial_02_Non_Blocking_Future_Result.scala
package com.allaboutscala.chap09


import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.{Failure, Success}

/**
 * Non blocking future result
 *
 * - With <b>Future.onComplete()</b> we are no longer blocking for the result from the Future but
 *   instead we will receive a callback for either a Success or a Failure.
 * - As such, we've also had to import scala.util.{Failure, Success}
 * - Surely though the Thread.sleep() is blocking our main thread so that we can see the
 *   asynchronous result from the future. In a real application, you will most certainly not use
 *   Thread.sleep() but instead "react" to the result returned by the future.
 */
object Tutorial_02_Non_Blocking_Future_Result extends App {

  println("Step 1: Define a function which returns a Future")
  def donutStock(donut: String): Future[Int] = Future {
    Thread.sleep(500) // assume some long running database operation
    println("checking donut stock")
    10
  }

  println("\nStep 2: Non blocking future result")
  // Instead of blocking our main program using Await.result(), we will make use of
  // Future.onComplete() callback to capture the result of a Future.
  donutStock("vanilla donut").onComplete {
    case Success(stock) => println(s"Stock for vanilla donut = $stock")
    case Failure(e) => println(s"Failed to find vanilla donut stock, exception = $e")
  }
  Thread.sleep(3000) // On laisse le temps à la future de compléter
}
